code/parse/sexp_container.h

```diff
@@ -28,29 +28,44 @@ enum class ContainerType : int {
 	SAVE_TO_PLAYER_FILE = 0x40000000,
 };
 
+using ContainerTypeInt = std::underlying_type<ContainerType>::type;
+
 inline constexpr bool operator&(ContainerType lhs, ContainerType rhs)
```

Comment:
Shouldn't this return `ContainerType` instead of `bool`?

Reply:
Strictly speaking, yes, it should.

However, if we return `ContainerType` here, intuitive expressions like `container.type & ContainerType::STRING_KEYS` then fail to compile, and there is no way to define implicit type conversion from an `enum class` to `bool` or to the underlying type `int` (which could then be implicitly converted to `bool` by the usual method)

And luckily, with the way *we* use `&` with `ContainerType` in practice, we expect a "yes"/"no" answer, hence our using it in conditional expressions throughout.

That said, if there's a way to return `ContainerType` form this function **and** get the above expression to compile without hacks, I'm open to considering it.

Reply:
FWIW, if for some reason, we actually need to get
```
ContainerType ct3 = ct1 & ct2;
```
we can still get it:
```
ContainerType ct3 = ct1;
ct3 &= ct2;
```

Reply:
> intuitive expressions like `container.type & ContainerType::STRING_KEYS` then fail to compile

How do you mean?  What is the full expression? because that is just part of an expression that doesn't use the return type at all.

> if for some reason, we actually do need

I think we would need that expression often enough that the workaround would be cumbersome and awkward.

And more significantly, the `operator|` *does* have the expected return type.  It violates the principle of least astonishment to have `&` and `|` behave differently.

Reply:
>How do you mean? What is the full expression?

Take a look at the `sexp_container` member functions `is_list()` and `is_map()` in `sexp_container.h` for examples of using `&` with `ContainerType` values to return a `bool` value, You can find examples of using `&` with `ContainerType` values as `if` conditions in the code changes to `missionsave.cpp` in PR #3449.

>I think we would need that expression often enough that the workaround would be cumbersome and awkward.

Offhand, I can't think of any instance of using `&` with a `sexp_container`'s `type` field (the original untyped `int` version) in the "full feature" PR #3398 where the intent is to use the result as a container type. Let me know if you find one, though.

>And more significantly, the `operator|` does have the expected return type. It violates the principle of least astonishment to have `&` and `|` behave differently.

Oh I agree that the current behavior with `operator&` returning `bool` is something of a convenient hack. It's a workaround for being unable to define implicit type conversion from an `enum class` to `bool`.

As such, I'll go looking for the least messy solution possible where `operator&`'s return type is changed to `ContainerType`  without excessively awkward rewordings of the usages of `&` I cited above.

FWIW the reason the "full feature" PR #3398 doesn't have this issue is that container type wasn't strongly typed/`enum class` then.

Reply:
I attempted to fulfill your request.

I am, frankly, unhappy with the result, which IMO is now more verbose, more error-prone (e.g., correctly using `==` vs. `!= with `ContainerType::NONE`), and considerably harder to read, but I don't have a better idea of how to provide the behavior you asked for while keeping the code working properly.

Reply:
This is better from a type safety perspective, and I don't think the readability suffers too much.  But you can improve it further by defining an `any` function and a `none` function, and then evaluate e.g. `any(ct & ContainerType::SOME_VALUE)`.  See this link for more info:
https://stackoverflow.com/questions/9874960/how-can-i-use-an-enum-class-in-a-boolean-context

Reply:
Ah, so that is how the type conversion from `enum class` to `bool` is performed. I saw that SO page earlier but didn't understand before what it was saying.

I concede that the resulting code does indeed look better and, as you said, now uses proper type conversion.